dynamic equality constraint

avikchaudhuri opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #99993

This diff adds support for dynamic equality constraints of the form `dynamic_dim(x, 0) == dynamic_dim(y, 1)`. The process of constraint discovery can already understand equality guards between dimensions and suggests such equality constraints, so this closes the loop on that. Correspondingly we now raise `ConstraintViolation` when we find that such a guard is added on a dynamic dimension and the user did not specify such a constraint. (NOTE: This is distinct from a dynamic dimension being guarded equal to a constant, which is already an error.)

Differential Revision: [D45279437](https://our.internmc.facebook.com/intern/diff/D45279437/)


cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

avikchaudhuri committed:
equalities



Differential Revision: [D45279437](https://our.internmc.facebook.com/intern/diff/D45279437/)

[ghstack-poisoned]

avikchaudhuri committed:
Update on "[WIP] dynamic equality constraint"


Differential Revision: [D45279437](https://our.internmc.facebook.com/intern/diff/D45279437/)

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

avikchaudhuri committed:
Update on "[WIP] dynamic equality constraint"


Differential Revision: [D45279437](https://our.internmc.facebook.com/intern/diff/D45279437/)

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

avikchaudhuri committed:
Update on "[WIP DO NOT REVIEW] dynamic equality constraint"


Differential Revision: [D45279437](https://our.internmc.facebook.com/intern/diff/D45279437/)

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

avikchaudhuri committed:
Update on "dynamic equality constraint"


This diff adds support for dynamic equality constraints of the form `dynamic_dim(x, 0) == dynamic_dim(y, 1)`. The process of constraint discovery can already understand equality guards between dimensions and suggests such equality constraints, so this closes the loop on that. Correspondingly we now raise `ConstraintViolation` when we find that such a guard is added on a dynamic dimension and the user did not specify such a constraint. (NOTE: This is distinct from a dynamic dimension being guarded equal to a constant, which is already an error.)

Differential Revision: [D45279437](https://our.internmc.facebook.com/intern/diff/D45279437/)


cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

guangy10 commented on test/dynamo/test_export.py:
```diff
@@ -2219,6 +2225,9 @@ def my_dyn_fn(a, b, c):
         constraints = [dynamic_dim(x, 0), dynamic_dim(y, 0), dynamic_dim(z, 0)]
+        with self.assertRaises(ConstraintViolationError):
+            torch._dynamo.export(my_dyn_fn, x, y, z, constraints=constraints)
+        constraints.append(dynamic_dim(z, 0) == dynamic_dim(x, 0))
```
I guess I will get more clarity as I keep reading the diff. Just to ask, dynamic dim equality means two dims must be both dynamic and have the same range values, correct?

avikchaudhuri replied:
Not just. They must be the same value. Usually this will be because there's a guard that forces them to be the same (e.g., they are compared equal, tensors with those dimensions are added together, etc.)

In contrast, 4 and 5 both satisfy the same range 3 < s < 6 (but are not equal).

guangy10 commented on torch/_dynamo/eval_frame.py:
```diff
@@ -627,26 +627,40 @@ def guard_export_print(guards):
 
 @dataclasses.dataclass
-class Constraint:
+class ConstraintTarget:
```
The new name looks a bit confusing. What does it mean?

avikchaudhuri replied:
It's the target of a constraint. The target is a (tensor, dimension) pair, basically. With this change I factored out the fields of `Constraint` into the target part and the actual constraint part (range), which also allows me to share the constraint part with another target.

guangy10 commented on torch/_dynamo/eval_frame.py:
```diff
@@ -671,6 +685,19 @@ def __bool__(self):
 
+    def __eq__(self, other):
+        constraint_range = StrictMinMaxConstraint(
+            vr=self.constraint_range.vr & other.constraint_range.vr,
```
hmm, why bitwise AND? what are the new lower/upper?

avikchaudhuri replied:
`ValueRange` overloads `&` to mean "intersection": `ValueRange(lower1, upper1) & ValueRange(lower2, upper2)` = ValueRange(max(lower1, lower2), min(upper1, upper2)`.

guangy10 commented on torch/_dynamo/guards.py:
```diff
@@ -472,10 +477,45 @@ def SHAPE_ENV(self, guard: Guard):
+        if output_graph.export_constraints:
+            source_pairs: List[Tuple[Source, Source]] = []
+            for constraint in output_graph.export_constraints:
+                source, *other_sources = get_sources(constraint.t_id, constraint.dim)
```
Given a tensor and a dim why would it have multiple sources?

avikchaudhuri replied:
Suppose that you want to export `def foo(x, y): ...` and your example inputs are `(t, t)`: in other words, you use the same tensor for both args. Then `x.size()[0]` and `y.size()[0]` are both sources of `(t, 0)`.

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -2115,9 +2146,8 @@ def produce_guards(
-            assert callable(msg_fn)
-            constraint_violations.append((constraint, msg_fn))
+        def record_constraint_violation(constraint, msg):
+            constraint_violations.append((constraint, lambda: msg))
```
moving the lambda in here is missing the point, the point is we need to delay the evaluation of msg

avikchaudhuri replied:
The way it was before, we were capturing variables by name that could change in value by the time the error is raised. This caused a lot of confusion for me. I had to meticulously rename captured variables in each message to ensure it accurately reflected the state when the lambda was created.

If we want to delay the evaluation of the message (what's so heavy about them?), we could pass in *args to the lambda.

ezyang replied:
Oh, yes, that's a legit bug!

The reason the messages need to be delayed is because the `hint()` call refers to `symbol_to_source`, which is still being built at the time we register constraints, in the loop here:

```
        def track_symint(source, val, constraint=None):
            if isinstance(val, SymInt):
                s = val.node.expr
        
                if isinstance(s, sympy.Symbol):
                    symbol_to_source[s].append(source)
                    if constraint is not None:
                        symbol_to_constraints[s].append(constraint)
                elif isinstance(-s, sympy.Symbol):
                    symbol_to_source[-s].append(NegateSource(source))
                else:
                    if constraint is not None:
                        # TODO: Maybe non-strict constraint shouldn't error
                        # here?  Check what happens in practice
                        def hint():
                            if s.free_symbols:
                                return (
                                    "Perhaps you meant to specify a constraint on {s.free_symbols}?" +
                                    "; ".join(
                                        f"{s0} bound by " + ", ".join(str(source0) for source0 in symbol_to_source[s0])
                                        for s0 in s.free_symbols
                                    )
                                )
                            else:
                                return "Did you really mean to mark this dimension as dynamic?"
            
                        record_constraint_violation(constraint, lambda: (
                            f"Could not validate constraint {constraint.render(source)} as "
                            f"{source.name()} is actually a non-atomic symbolic expression "
                            f"{s}.  {hint()}"
                        ))
```

I suppose you could index `symbol_to_source` FIRST, and then dispense with the lambdas. The main point is that some of the error messages rely on context which may not be entirely constructed at the time we produce the error.

avikchaudhuri replied:
I ended up limiting the use of capture to just the `hint` and using `functools.partial` to explicitly capture the variable `s` in the loop, which I believe was also getting captured incorrectly (it would have likely taken the last value out of the loop, given the hint would be called after the loop).

ezyang commented:
Didn't we agree we weren't going to add this?

avikchaudhuri commented:
I thought we were going to expand the constraint surface eventually, which is why we had an extensible format for constraints. See https://docs.google.com/document/d/1N8PxRfJ7IbshQiX5fsQFq8QHWh77SA8fi-Wojzug6ZA/edit#heading=h.iuvdfnjb9m4l

Equality constraints arise quite naturally, so it seems worthwhile to add. What's the downside?

ezyang commented:
> Equality constraints arise quite naturally, so it seems worthwhile to add. What's the downside?

The code is more complicated, and we have to maintain it going forward. Esp because we do have two major ShapeEnv changes coming down the pipe: multi-stage environment (for forward-backward split) and translation validation / de-sympification. The more complicated the preexisting code, the harder it will be to do these changes. We should make sure we are getting some value from each feature.

avikchaudhuri committed:
Update on "dynamic equality constraint"


This diff adds support for dynamic equality constraints of the form `dynamic_dim(x, 0) == dynamic_dim(y, 1)`. The process of constraint discovery can already understand equality guards between dimensions and suggests such equality constraints, so this closes the loop on that. Correspondingly we now raise `ConstraintViolation` when we find that such a guard is added on a dynamic dimension and the user did not specify such a constraint. (NOTE: This is distinct from a dynamic dimension being guarded equal to a constant, which is already an error.)

Differential Revision: [D45279437](https://our.internmc.facebook.com/intern/diff/D45279437/)


cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

avikchaudhuri commented:
Sorry, coming back to this diff after a few days. To your point about value vs. complexity: 
1. Does this really add a lot of complexity? The diff is IMO a super simple extension to what we already support, it doesn't require any new machinery. 
2. On value, let me try to make the case. I believe this is a necessary part of our soundness story with input constraints, which feels neither here nor there otherwise—either we don't raise constraint violations at all and tell people to just go test, or we try to be as complete as possible. I've seen equality constraints arise quite naturally, e.g., when inputs combine to feed into embeddings. Here's a walk through of a simple example.

```
def foo(x, y):
    return x + y
```

Of course you can choose to specialize everything, but let's think about what happens otherwise.
Suppose that I make `x.size()[0]` dynamic, then we already complain that it gets specialized to `y.size()[0]`. So we have to make `y.size()[0]` dynamic as well to get through.

```
x = torch.rand(5)
y = torch.rand(5)
gm, _ = dynamo.export(
    baz,
    x,
    y,
    constraints=[
        dynamic_dim(x, 0),
        dynamic_dim(y, 0),
    ],
    aten_graph=True,
    tracing_mode="symbolic",
    assume_static_by_default=True,
)
```
At this point, without this diff everything looks peachy, and our contract seems to suggest that we can accept any combination of input shapes. So we try:

```
gm(torch.rand(4), torch.rand(6))
```

And this crashes.

This diff will push such runtime errors to an early static error, which IMO is an important positive value.

tugsbayasgalan reviewed (approved):
LGTM!

avikchaudhuri committed:
Update on "dynamic equality constraint"


This diff adds support for dynamic equality constraints of the form `dynamic_dim(x, 0) == dynamic_dim(y, 1)`. The process of constraint discovery can already understand equality guards between dimensions and suggests such equality constraints, so this closes the loop on that. Correspondingly we now raise `ConstraintViolation` when we find that such a guard is added on a dynamic dimension and the user did not specify such a constraint. (NOTE: This is distinct from a dynamic dimension being guarded equal to a constant, which is already an error.)

Differential Revision: [D45279437](https://our.internmc.facebook.com/intern/diff/D45279437/)


cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

avikchaudhuri commented:
@pytorchbot merge
